File: archive/2023.6/Circle and Rectangle Overlapping/Main.cs
using System.Runtime.InteropServices;

namespace DailyCode;

public class Solution {
	// ReSharper disable once MemberCanBeMadeStatic.Global
#pragma warning disable CA1822
	public bool CheckOverlap(int radius, int xCenter, int yCenter, int x1, int y1, int x2, int y2) {
#pragma warning restore CA1822
		var xNearest = ((xCenter < x1) ? x1 : ((xCenter <= x2) ? xCenter : x2));
		var yNearest = ((yCenter < y1) ? y1 : ((yCenter <= y2) ? yCenter : y2));
		var distance = Math.Pow(xNearest - xCenter, 2) + Math.Pow(yNearest - yCenter, 2);
		return distance <= Math.Pow(radius, 2);
	}
}

File: archive/2023.6/Circle and Rectangle Overlapping/Main.cpp
#include <bits/stdc++.h>
#define VALUE_BETWEEN_RANGE(value, min, max, beforeMin, betweenValue, afterMax) ((value < min) ? beforeMin : ((value <= max) ? betweenValue : afterMax))
#define POW_DOUBLE(num) ((num) * (num))

using namespace std;

class Solution {
public:
	bool checkOverlap(int radius, int xCenter, int yCenter, int x1, int y1, int x2, int y2) {
		const int xNearest = VALUE_BETWEEN_RANGE(xCenter, x1, x2, x1, xCenter, x2);
		const int yNearest = VALUE_BETWEEN_RANGE(yCenter, y1, y2, y1, yCenter, y2);
		const int distance = POW_DOUBLE(xNearest - xCenter) + POW_DOUBLE(yNearest - yCenter);
		return distance <= POW_DOUBLE(radius);
	}
};
